File: LivePhotoDemo-OC/LivePhotoDemo-OC/ViewController.m
//
//  ViewController.m
//  LivePhotoDemo-OC
//
//  Created by Lcrnice on 2018/7/2.
//  Copyright © 2018年 Lcrnice. All rights reserved.
//

#import "ViewController.h"
#import "JPEG.h"
#import "QuickTimeMov.h"

@import Photos;
@import PhotosUI;
@import MobileCoreServices;

@interface ViewController () <UIImagePickerControllerDelegate, UINavigationControllerDelegate>

@property (weak, nonatomic) IBOutlet PHLivePhotoView *livePhotoView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self loadVideoWithVideoURL:[[NSBundle mainBundle] URLForResource:@"video" withExtension:@"m4v"]];
}

- (void)loadVideoWithVideoURL:(NSURL *)videoURL {
    self.livePhotoView.livePhoto = nil;
    AVURLAsset *asset = [AVURLAsset assetWithURL:videoURL];
    AVAssetImageGenerator *generator = [AVAssetImageGenerator assetImageGeneratorWithAsset:asset];
    generator.appliesPreferredTrackTransform = YES;
    NSValue *time = [NSValue valueWithCMTime:CMTimeMakeWithSeconds(CMTimeGetSeconds(asset.duration)/2, asset.duration.timescale)];
    [generator generateCGImagesAsynchronouslyForTimes:@[time] completionHandler:^(CMTime requestedTime, CGImageRef  _Nullable image, CMTime actualTime, AVAssetImageGeneratorResult result, NSError * _Nullable error) {
        if (image) {
            NSData *data = UIImagePNGRepresentation([UIImage imageWithCGImage:image]);
            NSArray <NSURL *>*urls = [[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask];
            NSURL *imageURL = [urls.firstObject URLByAppendingPathComponent:@"image.heic"];
            [data writeToURL:imageURL atomically:YES];
            
            NSString *image = imageURL.path;
            NSString *mov = videoURL.path;
            NSString *output = [ViewController livePath];
            NSString *assetIdentifier = [[NSUUID UUID] UUIDString];
            [[NSFileManager defaultManager] createDirectoryAtPath:output withIntermediateDirectories:YES attributes:nil error:&error];
            if (!error) {
                [[NSFileManager defaultManager] removeItemAtPath:[NSString stringWithFormat:@"%@/IMG.heic", output] error:&error];
                [[NSFileManager defaultManager] removeItemAtPath:[NSString stringWithFormat:@"%@/IMG.MOV", output] error:&error];
            }
            
            [[[JPEG alloc] initWithPath:image] writeToDest:[NSString stringWithFormat:@"%@/IMG.heic", output] withAssetIdentifier:assetIdentifier];
            [[[QuickTimeMov alloc] initWithPath:mov] writeToDestPath:[NSString stringWithFormat:@"%@/IMG.MOV", output] withAssetIdentifier:assetIdentifier];
            dispatch_sync(dispatch_get_main_queue(), ^{
                [PHLivePhoto requestLivePhotoWithResourceFileURLs:@[[NSURL fileURLWithPath:[[ViewController livePath] stringByAppendingString:@"/IMG.MOV"]], [NSURL fileURLWithPath:[[ViewController livePath] stringByAppendingString:@"/IMG.heic"]]] placeholderImage:nil targetSize:self.view.bounds.size contentMode:PHImageContentModeAspectFit resultHandler:^(PHLivePhoto * _Nullable livePhoto, NSDictionary * _Nonnull info) {
                    self.livePhotoView.livePhoto = livePhoto;
                    [self exportLivePhoto];
                }];
            });
        }
    }];
}

+ (NSString *)livePath {
    NSString *path = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).firstObject;
    path = [path stringByAppendingString:@"/"];
    return path;
}

- (void)exportLivePhoto {
    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^{
        PHAssetCreationRequest *createRequest = [PHAssetCreationRequest creationRequestForAsset];
        PHAssetResourceCreationOptions *options = [PHAssetResourceCreationOptions new];
        [createRequest addResourceWithType:PHAssetResourceTypePairedVideo fileURL:[NSURL fileURLWithPath:[[ViewController livePath] stringByAppendingString:@"/IMG.MOV"]] options:options];
        [createRequest addResourceWithType:PHAssetResourceTypePhoto fileURL:[NSURL fileURLWithPath:[[ViewController livePath] stringByAppendingString:@"/IMG.heic"]] options:options];
    } completionHandler:^(BOOL success, NSError * _Nullable error) {
        if (!success) {
            NSLog(@"%@", error.localizedDescription);
        }
    }];
}

- (IBAction)takePhoto:(UIBarButtonItem *)sender {
    UIImagePickerController *picker = [UIImagePickerController new];
    picker.delegate = self;
    picker.sourceType = UIImagePickerControllerSourceTypeCamera;
    picker.mediaTypes = @[(__bridge_transfer NSString *)kUTTypeMovie];
    [self presentViewController:picker animated:YES completion:nil];
}

#pragma mark - UIImagePickerControllerDelegate
- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info {
    [picker dismissViewControllerAnimated:YES completion:^{
        NSURL *url = info[UIImagePickerControllerMediaURL];
        [self loadVideoWithVideoURL:url];
    }];
}

@end

File: LivePhotoDemoSwift/ViewController.swift
//
//  ViewController.swift
//  LivePhotoDemoSwift
//
//  Created by Genady Okrain on 9/20/15.
//  Copyright © 2015 Genady Okrain. All rights reserved.
//

import UIKit
import Photos
import PhotosUI
import MobileCoreServices

//文件路径
struct FilePaths {
    static let documentsPath : AnyObject = NSSearchPathForDirectoriesInDomains(.cachesDirectory,.userDomainMask,true)[0] as AnyObject
    struct VidToLive {
        static var livePath = FilePaths.documentsPath.appending("/")
    }
}

class ViewController: UIViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    @IBOutlet weak var livePhotoView: PHLivePhotoView! {
        didSet {
            loadVideoWithVideoURL(Bundle.main.url(forResource: "video", withExtension: "MOV")!)
        }
    }
    
    //根据URL加载视频
    func loadVideoWithVideoURL(_ videoURL: URL) {
        livePhotoView.livePhoto = nil
        let asset = AVURLAsset(url: videoURL)
        let generator = AVAssetImageGenerator(asset: asset)
        generator.appliesPreferredTrackTransform = true
        let time = NSValue(time: CMTimeMakeWithSeconds(CMTimeGetSeconds(asset.duration)/2, asset.duration.timescale))
        generator.generateCGImagesAsynchronously(forTimes: [time]) { [weak self] _, image, _, _, _ in
            if let image = image, let data = UIImagePNGRepresentation(UIImage(cgImage: image)) {
                let urls = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
                let imageURL = urls[0].appendingPathComponent("image.heic")
                try? data.write(to: imageURL, options: [.atomic])
                
                let image = imageURL.path
                let mov = videoURL.path
                let output = FilePaths.VidToLive.livePath
                let assetIdentifier = UUID().uuidString
                let _ = try? FileManager.default.createDirectory(atPath: output, withIntermediateDirectories: true, attributes: nil)
                do {
                    try FileManager.default.removeItem(atPath: output + "/IMG.heic")
                    try FileManager.default.removeItem(atPath: output + "/IMG.MOV")
                } catch {
                    
                }
                JPEG(path: image).write(output + "/IMG.heic", assetIdentifier: assetIdentifier)
                QuickTimeMov(path: mov).write(output + "/IMG.MOV", assetIdentifier: assetIdentifier)
            
                //self?.livePhotoView.livePhoto = LPDLivePhoto.livePhotoWithImageURL(NSURL(fileURLWithPath: FilePaths.VidToLive.livePath.stringByAppendingString("/IMG.JPG")), videoURL: NSURL(fileURLWithPath: FilePaths.VidToLive.livePath.stringByAppendingString("/IMG.MOV")))
                //self?.exportLivePhoto()
                _ = DispatchQueue.main.sync {
                    PHLivePhoto.request(withResourceFileURLs: [ URL(fileURLWithPath: FilePaths.VidToLive.livePath + "/IMG.MOV"), URL(fileURLWithPath: FilePaths.VidToLive.livePath + "/IMG.heic")], placeholderImage: nil, targetSize: self!.view.bounds.size, contentMode: PHImageContentMode.aspectFit, resultHandler: { (livePhoto, info) -> Void in
                        self?.livePhotoView.livePhoto = livePhoto
                        self?.exportLivePhoto()
                    })
                }
            }
        }
    }
    
    //拍照
    @IBAction func takePhoto(_ sender: UIBarButtonItem) {
        let picker = UIImagePickerController()
        picker.delegate = self
        picker.sourceType = .camera
        picker.mediaTypes = [kUTTypeMovie as String]
        present(picker, animated: true, completion: nil)
    }
    //拍照回调
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
        picker.dismiss(animated: true) {
            if let url = info[UIImagePickerControllerMediaURL] as? URL {
                self.loadVideoWithVideoURL(url)
            }
        }
    }
    
    //保存Live图片到相册
    func exportLivePhoto () {
        PHPhotoLibrary.shared().performChanges({ () -> Void in
            let creationRequest = PHAssetCreationRequest.forAsset()
            let options = PHAssetResourceCreationOptions()
            
            creationRequest.addResource(with: PHAssetResourceType.pairedVideo, fileURL: URL(fileURLWithPath: FilePaths.VidToLive.livePath + "/IMG.MOV"), options: options)
            creationRequest.addResource(with: PHAssetResourceType.photo, fileURL: URL(fileURLWithPath: FilePaths.VidToLive.livePath + "/IMG.heic"), options: options)
        }, completionHandler: { (success, error) -> Void in
            if !success {
                DTLog((error?.localizedDescription)!)
            }
        })
    }
}

